SConstruct

```diff
@@ -0,0 +1,138 @@
+    ("LINK", "Linker"),
+    ("NUMCPUS", "Number of CPUs to use for build (0 means auto)", 0, None, int),
+    EnumVariable("BUILDTYPE", "Build type", "DEBUG", ["RELEASE", "DEBUG", "PERF"]),
+    BoolVariable("VERBOSE", "Show full build information", 0),
+    BoolVariable("WITH_GPROF", "Enable profiling", 0),
+    BoolVariable("BUILD_BINARIES", "Build binaries", 1),
+    PathVariable("PREFIX", "Installation target directory", "/usr/local/bin/", PathVariable.PathAccept),
+    PathVariable("DESTDIR", "The root directory to install into. Useful mainly for binary package building", "", PathVariable.PathAccept),
+)
+
+env = Environment(options = opts,
+                  tools = ['default'],
+                  ENV = os.environ)
+Help(opts.GenerateHelpText(env))
+
+# Copy environment variables
+if os.environ.has_key('CC'):
+    env["CC"] = os.getenv('CC')
+if os.environ.has_key('AS'):
+    env["AS"] = os.getenv('AS')
+if os.environ.has_key('LD'):
+    env["LINK"] = os.getenv('LD')
+if os.environ.has_key('CFLAGS'):
+    env.Append(CCFLAGS = SCons.Util.CLVar(os.environ['CFLAGS']))
+if os.environ.has_key('CPPFLAGS'):
+    env.Append(CPPFLAGS = SCons.Util.CLVar(os.environ['CPPFLAGS']))
+if os.environ.has_key('LDFLAGS'):
+    env.Append(LINKFLAGS = SCons.Util.CLVar(os.environ['LDFLAGS']))
+
+env.Append(CFLAGS = [ "-Wall", "-Werror", "-Wextra", "-O3", "-std=c99", "-I/usr/include/nspr"])
+
+if "WITH_GPROF" in env and env["WITH_GPROF"]:
+    print "With gprof"
+    env.Append(CFLAGS = [ "-pg" ])
+    env.Append(CPPFLAGS = [ "-pg" ])
+    env.Append(LINKFLAGS = [ "-pg" ])
+
+if env["BUILDTYPE"] == "DEBUG":
+    print "DEBUG MODE!"
+    env.Append(CPPFLAGS = [ "-g", "-DDEBUG"])
+elif env["BUILDTYPE"] == "RELEASE":
+    env.Append(CPPFLAGS = ["-DNDEBUG"])
+else:
+    print "Error BUILDTYPE must be RELEASE or DEBUG"
+    sys.exit(-1)
+
+if not env["VERBOSE"]:
+    env["CCCOMSTR"] = "Compiling $SOURCE"
+    env["SHCCCOMSTR"] = "Compiling $SOURCE"
+    env["ARCOMSTR"] = "Creating library $TARGET"
+    env["RANLIBCOMSTR"] = "Indexing library $TARGET"
+    env["LINKCOMSTR"] = "Linking $TARGET"
+
+def GetNumCPUs(env):
+    if env["NUMCPUS"] > 0:
+        return int(env["NUMCPUS"])
+    return 2*multiprocessing.cpu_count()
+
+env.SetOption('num_jobs', GetNumCPUs(env))
+
+# Configuration
+conf = env.Configure(custom_tests = { 'CheckPkgConfig' : CheckPkgConfig,
+                                      'CheckPkg' : CheckPkg,
+                                      'CheckPkgMinVersion' : CheckPkgMinVersion })
+
+if not conf.CheckCC():
+    print 'Your C compiler and/or environment is incorrectly configured.'
+    CheckFailed()
+
+conf.Finish()
+
+Export('env')
+
+# Set compile options for binaries
+env.Append(LIBS = ["mprio", "mpi", "nss3", "nspr4"], \
+  LIBPATH = ['#build/prio', "#build/mpi"])
+
+# Add header files
+env.Append(CPPPATH = ["#include", "#."])
+
+SConscript('prio/SConscript', variant_dir='build/prio')
+SConscript('mpi/SConscript', variant_dir='build/mpi')
```

Comment:
I'd like to see that we can use another copy of MPI here. That would allow us to use the copy in the Firefox tree. That should be relatively easy to do. We can keep the copy in the repo here to allow stand-alone builds but not vendor it into Firefox.

Reply:
@rhelmer Rob: Is this something that you can take care of on your side? I'm happy to modify the SConstruct file in libprio to accommodate using an external copy of MPI, but I'm not sure what the best way to do this is.